android/src/org/mozilla/firefox/vpn/InAppPurchase.kt

```diff
@@ -0,0 +1,293 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+@file:Suppress("KotlinJniMissingFunction")
+
+package org.mozilla.firefox.vpn
+
+import android.app.Activity
+import android.content.Context
+import android.util.Log
+import com.android.billingclient.api.BillingClient
+import com.android.billingclient.api.BillingClientStateListener
+import com.android.billingclient.api.BillingFlowParams
+import com.android.billingclient.api.BillingResult
+import com.android.billingclient.api.Purchase
+import com.android.billingclient.api.PurchasesUpdatedListener
+import com.android.billingclient.api.SkuDetails
+import com.android.billingclient.api.SkuDetailsParams
+import com.android.billingclient.api.SkuDetailsResponseListener
+import java.text.NumberFormat
+import java.util.Currency
+import kotlin.collections.ArrayList
+import kotlin.collections.HashMap
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class MozillaSubscriptionInfo(
+    val id: String,
+    val monthCount: Int
+)
+
+@Serializable
+data class MozillaSubscriptions(
+    val products: List<MozillaSubscriptionInfo>
+)
+
+@Serializable
+data class GooglePlaySubscriptionInfo(
+    val sku: String, // matches MozillaSubscriptionInfo.id
+    val totalPriceString: String,
+    val monthlyPriceString: String,
+    val monthlyPrice: Double,
+)
+
+@Serializable
+data class GooglePlaySubscriptions(
+    val products: ArrayList<GooglePlaySubscriptionInfo>
+)
+
+class InAppPurchase private constructor(ctx: Context) :
+    BillingClientStateListener, SkuDetailsResponseListener, PurchasesUpdatedListener {
+
+    /**
+     * SkuDetails and monthCounts by SKU
+     */
+    val skusWithSkuDetails = HashMap<String, SkuDetails>()
+    val skusWithMonthCount = HashMap<String, Int>()
+
+    /**
+     * The billingClient instance
+     */
+    private var billingClient = BillingClient.newBuilder(ctx)
+        .setListener(this)
+        .enablePendingPurchases() // Not used for subscriptions.
+        .build()
+
+    // Functions in AndroidIAPHandler
+    external fun onSkuDetailsReceived(subscriptionsDataJSONBlob: String)
+    external fun onNoPurchases()
+    external fun onPurchaseUpdated(purchaseDataJSONBlob: String)
+    external fun onSubscriptionFailed()
+
+    companion object {
+        private const val TAG = "InAppPurchase"
+
+        private var instance: InAppPurchase? = null
+
+        /**
+         * TODO - Discuss / get consult from basti
+         * Because we launch IAPHandler in mozillavpn, I believe all users will be forced
+         * to have Play and be logged in irrespective of whether they need to IAP.
+         * But the dependency on billing library may have this effect anyway.
+         */
```

Comment:
Given that we only publish with google play this should be fine. We already have other dependencies that break totally if the play store isn't present.